packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,146 @@
+        bytes memory signatures,
+        address msgSender
+    )
+        external
+    {
+        msgSender; // silence unused variable warning
+        require(msg.sender == address(SAFE), "LivenessGuard: only Safe can call this function");
+
+        // Cache the set of owners prior to execution.
+        // This will be used in the checkAfterExecution method.
+        address[] memory owners = SAFE.getOwners();
+        for (uint256 i = 0; i < owners.length; i++) {
+            ownersBefore.add(owners[i]);
+        }
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash({
+            to: to,
+            value: value,
+            data: data,
+            operation: operation,
+            safeTxGas: safeTxGas,
+            baseGas: baseGas,
+            gasPrice: gasPrice,
+            gasToken: gasToken,
+            refundReceiver: refundReceiver,
+            _nonce: Safe(payable(msg.sender)).nonce() - 1
+        });
+
+        uint256 threshold = SAFE.getThreshold();
+        address[] memory signers =
+            SafeSigners.getNSigners({ dataHash: txHash, signatures: signatures, requiredSignatures: threshold });
+
+        for (uint256 i = 0; i < signers.length; i++) {
+            lastLive[signers[i]] = block.timestamp;
+            emit OwnerRecorded(txHash, signers[i]);
+        }
+    }
+
+    /// @notice Update the lastLive mapping according to the set of owners before and after execution.
+    /// @dev Called by the Safe contract after the execution of a transaction.
+    ///      We use this post execution hook to compare the set of owners before and after.
+    ///      If the set of owners has changed then we:
+    ///      1. Add new owners to the lastLive mapping
+    ///      2. Delete removed owners from the lastLive mapping
+    function checkAfterExecution(bytes32, bool) external {
+        require(msg.sender == address(SAFE), "LivenessGuard: only Safe can call this function");
+
+        // Get the current set of owners
+        address[] memory ownersAfter = SAFE.getOwners();
+
+        // Iterate over the current owners, and remove one at a time from the ownersBefore set.
+        uint256 ownersAfterLength = ownersAfter.length;
+        for (uint256 i = 0; i < ownersAfterLength; i++) {
+            // If the value was present, remove() returns true.
+            address ownerAfter = ownersAfter[i];
+            if (ownersBefore.remove(ownerAfter) == false) {
+                // This address was not already an owner, add it to the lastLive mapping
+                lastLive[ownerAfter] = block.timestamp;
+            }
+        }
+        // Now iterate over the remaining ownersBefore entries. Any remaining addresses are no longer an owner, so we
+        // delete them from the lastLive mapping.
```

Comment:
you can cache `ownersBefore.length()`, not sure how much gas it saves because i don't know the implementation of `EnumerableSet`, also can use `j` but no need to because `i` above is not in scope